soci/soci_index.go

```diff
@@ -305,18 +305,42 @@ func NewIndexBuilder(contentStore content.Store, blobStore orascontent.Storage,
 	}, nil
 }
 
+// BuildAndPush builds a SOCI index, pushes and labels the artifacts, and returns the SOCI index.
+func (b *IndexBuilder) BuildAndPush(ctx context.Context, blobStore store.Store, img images.Image) (*IndexWithMetadata, error) {
+	sociIndexWithMetadata, done, err := b.Build(ctx, blobStore, img)
+	if err != nil {
+		return nil, err
+	}
+	defer done(ctx)
+
+	err = WriteSociIndex(ctx, sociIndexWithMetadata, blobStore, b.ArtifactsDb)
+	if err != nil {
+		return nil, err
+	}
+	return sociIndexWithMetadata, nil
+}
+
 // Build builds a soci index for `img` and return the index with metadata.
-func (b *IndexBuilder) Build(ctx context.Context, img images.Image) (*IndexWithMetadata, error) {
+// To prevent garbage collection of the zTOCs pushed by this command,
+// we obtain a lease from containerd and also return it.
+// It is the responsibility of the caller to terminate the lease.
+// Ideally the lease ends after labelling and pushing the index and related zTOCs.
+func (b *IndexBuilder) Build(ctx context.Context, blobStore store.Store, img images.Image) (*IndexWithMetadata, store.CleanupFunc, error) {
```

Comment:
This feels like a leaky abstraction. We create the bach here, but pass up the cleanup to the caller because there's more work to be done. Shouldn't the caller be the one to create the batch?

Reply:
Batch no longer gets returned